File: paleo_workbench/ui/pages/rich_text_preview_widget.py
from __future__ import annotations

from PySide6.QtWidgets import QTextBrowser, QTextEdit


class RichTextPreviewWidget(QTextBrowser):
    """Read-only rich-text renderer for Markdown/HTML.

    External network resources are blocked; local file:// images (relative to
    the document) are allowed so embedded figures render.
    """

    def __init__(self, parent=None):
        super().__init__(parent)
        self.setOpenExternalLinks(False)
        self.setReadOnly(True)

    def loadResource(self, resource_type, url):
        # Block non-file URLs (network). Allow file:// for local images.
        if url.scheme() not in ("", "file"):
            return None
        return super().loadResource(resource_type, url)

    def load_html(self, html: str) -> None:
        self.setHtml(html)

    def apply_settings(self, settings) -> None:
        font = self.font()
        font.setPointSize(settings.font_size)
        self.setFont(font)
        self.setLineWrapMode(
            QTextEdit.LineWrapMode.WidgetWidth
            if settings.wrap_text
            else QTextEdit.LineWrapMode.NoWrap
        )

File: libs/ui_pages_preview/src/qt/rich_text_preview_widget.cpp
#include <pwb/ui_pages_preview/qt/rich_text_preview_widget.hpp>

#include <pwb/ui_pages_preview/preview_settings.hpp>
#include <pwb/ui_pages_preview/url_filter.hpp>

namespace pwb::ui_pages_preview {

RichTextPreviewWidget::RichTextPreviewWidget(QWidget* parent)
    : QTextBrowser(parent) {
    setOpenExternalLinks(false);
    setReadOnly(true);
}

QVariant RichTextPreviewWidget::loadResource(int type, const QUrl& name) {
    if (!resource_scheme_allowed(name.scheme().toStdString())) {
        return {};
    }
    return QTextBrowser::loadResource(type, name);
}

void RichTextPreviewWidget::apply_settings(const PreviewSettings& settings) {
    QFont f = font();
    f.setPointSize(settings.font_size);
    setFont(f);
    setLineWrapMode(settings.wrap_text ? QTextEdit::WidgetWidth
                                       : QTextEdit::NoWrap);
}

}  // namespace pwb::ui_pages_preview
